Remove enumerable runtime checking by default

aisamanra opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->


### Motivation
This adds an extra parameter to `valid?` that defaults to `false` which governs whether it recurses into generic collection types for runtime type-checking. This allows us to avoid it by default (avoiding a big source of slowdown) but still do it if we need to, and opens up the possibility of future features which can "opt-in" to recursive checking for extra safety.

This also adds the `deep` parameter to `assert_type!`; methods which rely on "deep" runtime checking can retain their old externally-visible runtime behavior by using something like this:

```
sig {params(xs: T::Hash[String, T::Array[Integer]]).void}
def foo(xs)
  assert_type!(xs, T::Hash[String, T::Array[Integer]], deep: true)
  # ...
end
```

A standing question is whether we want to enable this by default for sigs using e.g. a construct like `.checked(:deep)`. Also, should the `deep` parameter be called something like `recursive`?

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->
Updated tests to include both recursive and non-recursive variants of the same behavior. TODO: test over pay-server.

See included automated tests.

jez commented on gems/sorbet-runtime/lib/types/private/types/not_typed.rb:
```diff
@@ -12,7 +12,7 @@ def name
 
   # @override Base
-  def valid?(obj)
+  def valid?(obj, deep=false)
```
Why is this positional while the others are keyword? I generally prefer that optional boolean arguments are keyword arguments.

jez replied:
On second thought, seeing how this is used throughout the typed_enumerable code below, it seems really easy to forget to forward through the `deep` parameter. Can we make this parameter required, and change pay-server?

jez commented on gems/sorbet-runtime/lib/types/types/typed_array.rb:
```diff
@@ -13,7 +13,7 @@ def underlying_class
     # @override Base
-    def valid?(obj)
+    def valid?(obj, deep=false)
       obj.is_a?(Array) && super
```
ooc, we could short circuit all of these is `deep` is `true`, right? Why have you chosen to unconditionally call `super`?

jez commented:
@aisamanra Can you remind me what the motivation is here?

aisamanra commented:
This now has a `recursively_valid?` (instead of `valid?` with an argument) and I've removed the `deep:` keyword argument in favor of an entirely separate top-level helper method: this will make it trivial to delete if we ever want to do this.

jez commented on gems/sorbet-runtime/lib/types/_types.rb:
```diff
@@ -155,6 +155,14 @@ def self.assert_type!(value, type, checked: true)
 
+  # Tells the typechecker to ensure that `value` is of type `type` (if not, the typechecker will
+  # fail). Use this for debugging typechecking errors, or to ensure that type information is
+  # statically known and being checked appropriately. If `checked` is true, raises an exception at
+  # runtime if the value doesn't match the type.
```
This comment looks out of date?

aisamanra replied:
Comment updated to be correct now. I've also added a bit to the docs, since I had forgotten to do that.

jez commented on gems/sorbet-runtime/lib/types/_types.rb:
```diff
@@ -155,6 +155,14 @@ def self.assert_type!(value, type, checked: true)
 
+  # Dynamically confirm that `value` is recursively a valid value of
+  # type `type`, including recursively through collections. Note that
+  # in some cases this runtime check can be very expensive, especially
+  # with large collections of objects.
```
```suggestion
  # in some cases this runtime check can be very expensive, especially
  # with large collections of objects.
  #
  # This doesn't have any connection to static type refinements at all.
  # If you additionally want the type to be cast statically, use `T.let` or `T.cast`.
```

jez commented on gems/sorbet-runtime/lib/types/types/typed_array.rb:
```diff
@@ -13,10 +13,15 @@ def underlying_class
     # @override Base
-    def valid?(obj)
+    def recursively_valid?(obj)
       obj.is_a?(Array) && super
```
```suggestion
      valid?(obj) && super
```

jez replied:
^ This might be a change you want to make in a few other places too (just for consistency)

jez commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -54,10 +58,10 @@ def valid?(obj)
       when Set
         obj.each do |item|
-          return false unless @type.valid?(item)
+          return false unless @type.recursively_valid?(item)
```
I like this a lot better with the `recursively_valid?` naming / boolean blindness issues.

jez commented on rbi/sorbet/t.rbi:
```diff
@@ -32,6 +32,9 @@ module T
   def self.assert_type!(value, type, checked: true); end
 
+  sig {params(value: T.untyped, type: T.untyped).returns(BasicObject)}
+  def self.check_type_recursive!(value, type); end
```
Do we want to put this in `T::Private`? Do we intend this to be a public API?

jez commented on gems/sorbet-runtime/test/types/props/decorator.rb:
```diff
@@ -173,11 +173,7 @@ class InnerStruct < T::Struct
-
-      assert_raises(TypeError) do
-        StructHash.new(the_hash: {'foo' => {}})
-      end
```
Was this deleted just because it's identical to below? Is it worth keeping the `.new` call around and just deleting the `assert_raises`?

jez commented on gems/sorbet-runtime/test/types/casts.rb:
```diff
@@ -44,5 +44,27 @@ class CastsTest < Critic::Unit::UnitTest
+
+      it "does do recursive type-checking of arrays with `check_type_recursive!`" do
+        assert_raises(TypeError) do
+          T.check_type_recursive!([1], T::Array[String])
```
Yeah seeing this twice, I'm leaning towards at least putting this in `T::Utils` or something, and not putting it on `T` directly, because I think we don't really want to encourage people to depend on this accidentally.

aisamanra replied:
I think that for _now_ I'd advocate for `T::Utils`, but it depends on how heavily we think we'll remove this entirely in the future. I think it makes some sense to keep it around, but if we're otherwise pretty sure it'll be removed at some point, then maybe `T::Private` is the right place for it?

jez commented on gems/sorbet-runtime/lib/types/types/fixed_hash.rb:
```diff
@@ -16,6 +16,21 @@ def name
+      end
+
+      true
+    end
```
Feel free to push back here. Do you think this method could be improved by implementing it something like

```ruby
valid?(obj) && <logic for recursively_valid>
```

?

(motivation: looks like a lot of duplication with `valid?`)

aisamanra replied:
I was going to push back for performance concerns (since it'll re-traverse the hash twice, once for `valid?` and once for `recursively_valid?`) but on reflection, if you're using `recursively_valid?`, then you're already throwing perf out the window, so I'll do this refactor.

djudd-stripe commented on gems/sorbet-runtime/lib/types/types/fixed_array.rb:
```diff
@@ -17,6 +17,12 @@ def name
+    # @override Base
+    def recursively_valid?(obj)
+      obj.is_a?(Array) && obj.length == @types.length &&
+        obj.zip(@types).all? {|item, type| type.recursively_valid?(item)}
```
Might want to rebase after https://github.com/sorbet/sorbet/pull/3302 is merged

djudd-stripe commented on gems/sorbet-runtime/lib/types/types/fixed_hash.rb:
```diff
@@ -16,6 +16,13 @@ def name
     end
 
+    # @override Base
+    def recursively_valid?(obj)
```
Ditto, might want to rebase after https://github.com/sorbet/sorbet/pull/3302

djudd-stripe commented on gems/sorbet-runtime/lib/types/types/union.rb:
```diff
@@ -42,6 +42,11 @@ def name
     end
 
+    # @override Base
+    def recursively_valid?(obj)
```
Seems like the non-enumerable variants of `recursively_valid?` don't have any test coverage - am I missing something?

aisamanra replied:
You're right—I've gone ahead and updated the `types` test now to test not only both code paths but ensure that the returned error messages are identical in every test where we'd expect them to be the same.

aisamanra commented:
@jez I've made a few changes and so for thoroughness I figured I'd request a re-review. The changes since the last one are
1. setters and constructors to `T::Struct` and friends are now type-checked
2. the tests are now more thorough, and for everything where `valid?` and `recursively_valid?` are expected to be the same it'll test both and also test that the two parameters are identical.

jez commented:
> setters and constructors to T::Struct and friends are now type-checked

I assume this is at the request of kv-client, right? would you mind adding a note above the `recursively_valid?` calls mentioning something that it was an explicit decision to do that?

jez commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -311,10 +323,16 @@ def subtype_of?(other)
+      it 'fails if an element of the array is the wrong type under deep checking' do
+        type = T::Array[Integer]
+        value = [true]
+        msg = type.error_message_for_obj_recursive(value)
```
👍

djudd-stripe commented on gems/sorbet-runtime/lib/types/types/typed_range.rb:
```diff
@@ -15,10 +15,15 @@ def name
 
     # @override Base
-    def valid?(obj)
+    def recurisvely_valid?(obj)
```
Typo?

aisamanra replied:
Oof, yes. …apparently not only do we not test that, we must not use it anywhere in pay-server, either.
